src/librbd/librbd.cc

```diff
@@ -1013,8 +1016,18 @@ namespace librbd {
   {
     ImageCtx *ictx = (ImageCtx *)ctx;
     tracepoint(librbd, snap_remove_enter, ictx, ictx->name.c_str(), ictx->snap_name.c_str(), ictx->read_only, snap_name);
+    cls::rbd::SnapshotNamespace snap_namespace;
+    int r = get_snap_namespace(ictx, snap_name, &snap_namespace);
+    if (r < 0) {
+      tracepoint(librbd, snap_remove_exit, r);
+      return r;
+    }
+    if (!(snap_namespace == static_cast<cls::rbd::SnapshotNamespace>(cls::rbd::UserSnapshotNamespace()))) {
```

Comment:
Sorry, read that backwards. I think it would be cleaner to just use:

```if (boost::get<cls::rbd::UserSnapshotNamespace>(&snap_namespace) == nullptr)```

... also, this should probably be moved into the ```snap_remove``` method so that both the C and C++ API follow the same logic.